pkg/chunked/compression.go

```diff
@@ -0,0 +1,513 @@
+package chunked
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"time"
+
+	"github.com/containers/storage/pkg/ioutils"
+	"github.com/klauspost/compress/zstd"
+	digest "github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+	"github.com/vbatts/tar-split/archive/tar"
+)
+
+type zstdTOC struct {
+	Version int                `json:"version"`
+	Entries []zstdFileMetadata `json:"entries"`
+}
+
+type zstdFileMetadata struct {
+	Type       string            `json:"type"`
+	Name       string            `json:"name"`
+	Linkname   string            `json:"linkName,omitempty"`
+	Mode       int64             `json:"mode,omitempty"`
+	Size       int64             `json:"size"`
+	UID        int               `json:"uid"`
+	GID        int               `json:"gid"`
+	ModTime    time.Time         `json:"modtime"`
+	AccessTime time.Time         `json:"accesstime"`
+	ChangeTime time.Time         `json:"changetime"`
+	Devmajor   int64             `json:"devMajor"`
+	Devminor   int64             `json:"devMinor"`
+	Xattrs     map[string]string `json:"xattrs,omitempty"`
+	Digest     string            `json:"digest,omitempty"`
+	Offset     int64             `json:"offset,omitempty"`
+	EndOffset  int64             `json:"endOffset,omitempty"`
+
+	// Currently chunking is not supported.
+	ChunkSize   int64  `json:"chunkSize,omitempty"`
+	ChunkOffset int64  `json:"chunkOffset,omitempty"`
+	ChunkDigest string `json:"chunkDigest,omitempty"`
+}
+
+const (
+	TypeReg     = "reg"
+	TypeChunk   = "chunk"
+	TypeLink    = "hardlink"
+	TypeChar    = "char"
+	TypeBlock   = "block"
+	TypeDir     = "dir"
+	TypeFifo    = "fifo"
+	TypeSymlink = "symlink"
+)
+
+var tarTypes = map[byte]string{
+	tar.TypeReg:     TypeReg,
+	tar.TypeRegA:    TypeReg,
+	tar.TypeLink:    TypeLink,
+	tar.TypeChar:    TypeChar,
+	tar.TypeBlock:   TypeBlock,
+	tar.TypeDir:     TypeDir,
+	tar.TypeFifo:    TypeFifo,
+	tar.TypeSymlink: TypeSymlink,
+}
+
+var typesToTar = map[string]byte{
+	TypeReg:     tar.TypeReg,
+	TypeLink:    tar.TypeLink,
+	TypeChar:    tar.TypeChar,
+	TypeBlock:   tar.TypeBlock,
+	TypeDir:     tar.TypeDir,
+	TypeFifo:    tar.TypeFifo,
+	TypeSymlink: tar.TypeSymlink,
+}
+
+func getType(t byte) (string, error) {
+	r, found := tarTypes[t]
+	if !found {
+		return "", fmt.Errorf("unknown tarball type: %v", t)
+	}
+	return r, nil
+}
+
+func typeToTarType(t string) (byte, error) {
+	r, found := typesToTar[t]
+	if !found {
+		return 0, fmt.Errorf("unknown type: %v", t)
+	}
+	return r, nil
+}
+
+const (
+	manifestChecksumKey = "io.containers.zstd-chunked.manifest-checksum"
```

Comment:
Could we store uncompressed digest of TOC manifest to this or have another annotation for it? (https://github.com/containerd/stargz-snapshotter/pull/293#issuecomment-815657316)

Reply:
sure, I can add more metadata if it helps you.

How will it be different than `io.containers.zstd-chunked.manifest-checksum`?

Reply:
This is for compatibility with estargz. Our snapshotter verifies TOC manifest using its uncompressed digest so it would be great if zstd:chunked contains an annotation for it. But current `io.containers.zstd-chunked.manifest-checksum` seems to contain the digest of zstd-compressed TOC manifest, right?

Except for this annotation issue, zstdchunked image (`ghcr.io/ktock/python:3.9-zstdchunked`) created using `c/storage/pkg/chunked` works fine with containerd + stargz snapshotter!

Reply:
In terms of the annotation name, it would be great if it's prefixed by `containerd.io/snapshot/` because it allows stargz snapshotter to support zstd:chunked without adding special patches to `containerd/containerd` project.